readme.txt
=== Connect Paid Memberships Pro to Discord ===
Contributors: expresstechsoftware, webbdeveloper, sunnysoni, strangerstudios 
Tags: Discord, Talk, Video Chat, Hang Out, Friends, Memberships, discord role management
Donate link: https://paypal.me/supportets
Author URI: https://www.expresstechsoftwares.com
Author: ExpressTech Software Solutions Pvt. Ltd.
Requires at least: 4.7
Tested up to: 6.3
Requires PHP: 7.0
Stable tag: 1.2.11
License: GPLv2
License URI: https://www.gnu.org/licenses/gpl-2.0.html

This add-on enables connecting your PMPRO enabled website to your discord server. Now you can add/remove PMPRO members directly to your discord server roles, assign roles according to your member levels, unassign roles when they expire, change role when they change membership.
 
== Description ==
= This PMPRO Discord AddOn enables connecting your PaidMebershipPro members to connect to your discord online community, with the server roles assigned to members as per their membership level. =

Very simple setup and intutive User interface to Manage Member Role inside Discord.
- Please always contact support if you are facing issues in setup or see any bug.
- If you did like the plugin, kindly support us in doing free excellent work our PayPal email is: business@expresstechsoftwares.com

[youtube https://youtu.be/v7lxB_Bvlv4]

This plugin provides the following features: 
1) Allow any member to connect their discord account with their PaidMebershipPro membership account. 
2) Members will be assigned roles in discord as per their membership level.
3) Members roles can be changed/remove from the admin of the site.
4) Members roles will be updated when membership expires.
5) Members roles will be updated when membership cancelled.
6) Admin can decide what default role to be given to all members upon connecting their discord to their membership account.
7) Admin can decide if membership should stay in their discord server when membership expires or cancelled.
8) Admin can decide what default role to be assigned when membership cancelled or expire.
9) Admin can change role by changing the membership by editng user insider WP Manage user.
10) Send a Direct message to discord members when their membership has expired. (Only work when allow none member is set to YES and Direct Message advanced setting is set ENABLED)
11) Send a Direct message to discord members when their membership is cancelled. (Only work when allow none member is set to YES and Direct Message advanced setting is set ENABLED)
12) Send membership expiration warnings Direct Message when membership is about to expire (Default 7 days before)
13) Short code [discord_connect_button] can be used on any page to display connect/disconnect button.
14) Using the shortcode [discord_connect_button] on any page, anyone can join the website discord server by authentication via member discord account. New members will get `default` role if selected in the setting.
15) Button styling feature under the plugin settings.
16) Support of Paid Memberships Pro - Cancel on Next Payment Date. So the member role wont get removed immediately upon cancel.
17) Support for forced discord authentication before checkout.
18) Hide the connect button using the simple filter: ets_pmpro_show_connect_button_on_profile
`add_filter('ets_pmpro_show_connect_button_on_profile', '__return_false' );`
Adding above code line in functions.php of theme or using code snippet plugin.

[View all Screenshots](https://www.expresstechsoftwares.com/pmpro-official-discord-add-on/)

== Installation ==

= Download, Install and Activate! =
1. Go to Plugins > Add New to find and install PMPRO discord Addon.
2. Or, download the latest version of the plugin, then go to Plugins > Add New and click the "Upload Plugin" button to upload your .zip file.
3. Activate the plugin.

= Complete the Initial Plugin Setup =
Go to Memberships > Discord Settings in the WordPress admin to begin setup. Our [Initial Setup Tutorial Video ](https://youtu.be/v7lxB_Bvlv4) will show you how to configure

= More Installation and Setup Documentation =
* [Installation Video](https://youtu.be/v7lxB_Bvlv4)
* [Installation Docs](https://www.expresstechsoftwares.com/step-by-step-documentation-guide-on-how-to-connect-pmpro-and-discord-server-using-discord-addon/)

== Checkout Our Other Plugins ==
1. [Connect MemberPress and Discord](https://wordpress.org/plugins/expresstechsoftwares-memberpress-discord-add-on/)
2. [connect Learnpress and Discord](https://wordpress.org/plugins/connect-learnpress-discord-add-on/)
3. [Connect LearnDash and Discord](https://wordpress.org/plugins/connect-learndash-and-discord/)
4. [Connect GamiPress and Discord](https://wordpress.org/plugins/connect-gamipress-and-discord/)
5. [Connect LifterLMS and Discord](https://wordpress.org/plugins/connect-lifterlms-to-discord/)
6. [Webhook For WCFM Vendors](https://wordpress.org/plugins/webhook-for-wcfm-vendors/)
7. [Inspect HTTP Requests](https://wordpress.org/plugins/inspect-http-requests/)
8. [Product Questions & Answers for WooCommerce](https://wordpress.org/plugins/product-questions-answers-for-woocommerce/)
9. [Connect Ultimate Member and Discord](https://wordpress.org/plugins/ultimate-member-discord-add-on/)
10. [Connect Eduma Theme and Discord](https://wordpress.org/plugins/connect-eduma-theme-to-discord/)
11. [Connect BadgeOS and Discord](https://wordpress.org/plugins/connect-badgeos-to-discord/)


== Frequently Asked Questions ==
= I need help installing, configuring, or customizing the plugin. =
Please visit [our support site at https://www.expresstechsoftwares.com/step-by-step-documentation-guide-on-how-to-connect-pmpro-and-discord-server-using-discord-addon/](https://www.expresstechsoftwares.com/step-by-step-documentation-guide-on-how-to-connect-pmpro-and-discord-server-using-discord-addon/) for more documentation and our support forums.
= I'm getting an error in error Log 'Missing Access'
Please make sure your bot role has the highest priority among all other roles in your discord server roles settings. please watch video on youtube how to do it. [Installation Video](https://youtu.be/v7lxB_Bvlv4?t=363)
= Role Settings is not appearing.
1. Clear browser cache, to uninstall and install again.
2. Try the disabling cache
3. Try Disabling other plugins, there may be any conflict with another plugin.
= Members are not being added spontaneously.
1. Due to the nature of Discord API, we have to use schedules to precisely control API calls, that why actions are delayed. 
= Member roles are not being assigned spontaneously.
1. Due to the nature of Discord API, we have to use schedules to precisely control API calls, that why actions are delayed. 
= Some members are not getting their role and there is no error in the log.
1. Sometimes discord API behaves weirdly, It is suggested to TRY again OR use another discord account.
= After expiry or member cancellation the roles are not being removed
1. It is seen in discord API that it return SUCCESS but does not work sometimes. It is suggested to manually adjust roles via PMPPRO->Members table.

== Screenshots ==
1. Install and activate the plugin and view the discord settings page inside Memberships
2. Map Discord roles and PMPRO levels.
3. Advanced settings.
4. Spot the Connect to Discord on your profile page.
5. The appearance and text of the connect / disconnect button.


CHANGELOG.txt
= 1.2.12 =
- Update the dismissible notification asking for review

= 1.2.11 =
- Add dismissible notification asking for review

= 1.2.10 =
- Update WordPress version 6.3

= 1.2.9 =
- Fixed #GH 319

= 1.2.8 =
- Add Opt-out
- Fixed #GH 310
- Fixed #GH 315
- Fixed #GH 316

= 1.2.7 =
- Add link to support forum

= 1.2.6 =
- Add visual clues to demonstrate the error connect to bot cases.

= 1.2.5 =
- Fix 10004 error

= 1.2.4 =
- Hide the connect button using the simple filter: ets_pmpro_show_connect_button_on_profile

= 1.2.3 =
- Added filter #GH 304

= 1.2.2 =
- Fixed a BUG GH#298

= 1.2.1 =
- Add linebreak feature in DM messages.
- Added color on roles under connect and disconnect button.

= 1.2.0 =
- Fixed the BUG GH#294, Fatal error Call to undefined function is_plugin_active()

= 1.1.9 =
- Added support for Discord API v10.

= 1.1.8 =
- Added support for Paid Memberships Pro - Cancel on Next Payment Date

= 1.1.7 =
- Member not getting added to guild, GH#282

= 1.1.6 =
- Fixed a BUG, where member do disconnect and can have role for infinite time

= 1.1.5 =
- Fixed a BUG which caused Connect and Disconnect buttons to disappear.

= 1.1.4 =
- Fixed a security issue, where anyone can trigger connect to bot.

= 1.1.3 =
- Added button styling feature under the plugin settings.
- Added A button under Logs tabs to see pending scheduled actions.
- Bot token input now secured.
- Removed the use of $_SERVER
- Remove refresh token when member do disconnect

= 1.1.2 =
- In this version, we fixed a bug related to the advanced setting of a "Login with Discord"

= 1.1.1 =
* Fixed the issue of shortcode [discord_connect_button] is breaking the site because the member is not logged-in
* Added a new feature, using the shortcode [discord_connect_button] on any page, anyone can join the website discord server.

= 1.1.0 =
* Fixed saving shortcode [discord_connect_button] error while saving inside gutenberg editor.
* Feature one discord role can be mapped to many PMPRO Levels.

= 1.0.6 =
* Fixed CSS/JS inclusion only where needed.

= 1.0.5 =
* Fixed JS error related to the TABS plugin.

= 1.0.4 =
* Fixed CSS conflict with other discord plugins.

= 1.0.3 =
* Fixed assets folder 

= 1.0.2 =
* Added shortcode [discord_connect_button]
* Secure the log file by adding UUID4 prefix in the file name during plugin activation.

assets/js/ets-pmpro-add-discord-script.js
jQuery(document).ready(function ($) {
	if (etsPmproParams.is_admin) {
		if(window.location.href.indexOf("ets_pmpro_") == -1) {
			jQuery("#skeletabsTab1").trigger("click");
		}
		/*Load all roles from discord server*/
		$.ajax({
			type: "POST",
			dataType: "JSON",
			url: etsPmproParams.admin_ajax,
			data: { 'action': 'ets_pmpro_discord_load_discord_roles', 'ets_discord_nonce': etsPmproParams.ets_discord_nonce, },
			beforeSend: function () {
				$(".pmpro-discord-roles .spinner").addClass("is-active");
				$(".initialtab.spinner").addClass("is-active");
			},
			success: function (response) {
				if (response != null && response.hasOwnProperty('code') && response.code == 50001 && response.message == 'Missing Access') {
					$(".pmpro-btn-connect-to-bot").show();
				}else if ( response.code === 10004 && response.message == 'Unknown Guild' ){
					$(".pmpro-btn-connect-to-bot").show().after('<p><b>The server ID is wrong or you did not connect the Bot.</b></p>');
				} else if ( response.code === 0 && response.message == '401: Unauthorized' ){
					$("#pmpro-connect-discord-bot").show().html("Error: Unauthorized - The Bot Token is wrong").addClass('error-bk');
				} else if (response == null || response.message == '401: Unauthorized' || response.hasOwnProperty('code') || response == 0) {
					$("#pmpro-connect-discord-bot").show().html("Please check all details are correct Or Click to connect the Bot").addClass('error-bk');
				} else {
					if ($('.ets-tabs button[data-identity="level-mapping"]').length) {
						$('.ets-tabs button[data-identity="level-mapping"]').show();
					}
					$("#pmpro-connect-discord-bot").show().html("Bot Connected <i class='fab fa-discord'></i>").addClass('not-active');

					var activeTab = localStorage.getItem('activeTab');
					if ($('.ets-tabs button[data-identity="level-mapping"]').length == 0 && activeTab == 'level-mapping') {
						$('.ets-tabs button[data-identity="settings"]').trigger('click');
					}
					$.each(response, function (key, val) {
						var isbot = false;
						if (val.hasOwnProperty('tags')) {
							if (val.tags.hasOwnProperty('bot_id')) {
								isbot = true;
							}
						}

						if (key != 'previous_mapping' && isbot == false && val.name != '@everyone') {
							$('.pmpro-discord-roles').append('<div class="makeMeDraggable" style="background-color:#'+val.color.toString(16)+'" data-pmpro_role_id="' + val.id + '" >' + val.name + '</div>');
							$('#pmpro-defaultRole').append('<option value="' + val.id + '" >' + val.name + '</option>');
							makeDrag($('.makeMeDraggable'));
						}
					});
					var defaultRole = $('#selected_default_role').val();
					if (defaultRole) {
						$('#pmpro-defaultRole option[value=' + defaultRole + ']').prop('selected', true);
					}

					if (response.previous_mapping) {
						var mapjson = response.previous_mapping;
					} else {
						var mapjson = localStorage.getItem('pmpro_mappingjson');
					}

					$("#pmpro_maaping_json_val").html(mapjson);
					$.each(JSON.parse(mapjson), function (key, val) {
						var arrayofkey = key.split('id_');
						var preclone = $('*[data-pmpro_role_id="' + val + '"]').clone();
						if(preclone.length>1){
							preclone.slice(1).hide();
						}
						//$('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').append(preclone).attr('data-drop-pmpro_role_id', val).find('span').css({ 'order': '2' });
						if (jQuery('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').find('*[data-pmpro_role_id="' + val + '"]').length == 0) {
							$('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').append(preclone).attr('data-drop-pmpro_role_id', val).find('span').css({ 'order': '2' });
						}
						if ($('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').find('.makeMeDraggable').length >= 1) {
							$('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').droppable("destroy");
						}
						// if (jQuery('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').find('.makeMeDraggable').length >= 1) {
						// 	$('*[data-pmpro_level_id="' + arrayofkey[1] + '"]').droppable("destroy");
						// }
						preclone.css({ 'width': '100%', 'left': '0', 'top': '0', 'margin-bottom': '0px', 'order': '1' }).attr('data-pmpro_level_id', arrayofkey[1]);
						makeDrag(preclone);
					});
				}

			},
			error: function (response) {
				$("#pmpro-connect-discord-bot").show().html("Error: Please check all details are correct").addClass('error-bk');
				console.error(response);
			},
			complete: function () {
				$(".pmpro-discord-roles .spinner").removeClass("is-active").css({ "float": "right" });
				$("#skeletabsTab1 .spinner").removeClass("is-active").css({ "float": "right", "display": "none" });
			}
		});


		/*Clear log log call-back*/
		$('#pmpro-clrbtn').click(function (e) {
			e.preventDefault();
			$.ajax({
				url: etsPmproParams.admin_ajax,
				type: "POST",
				data: { 'action': 'ets_pmpro_discord_clear_logs', 'ets_discord_nonce': etsPmproParams.ets_discord_nonce, },
				beforeSend: function () {
					$(".clr-log.spinner").addClass("is-active").show();
				},
				success: function (data) {
					if (data.error) {
						// handle the error
						alert(data.error.msg);
					} else {
						$('.error-log').html("Clear logs Sucesssfully !");
					}
				},
				error: function (response) {
					console.error(response);
				},
				complete: function () {
					$(".clr-log.spinner").removeClass("is-active").hide();
				}
			});
		});

		/*Flush settings from local storage*/
		$("#revertMapping").on('click', function () {
			localStorage.removeItem('pmpro_mapArray');
			localStorage.removeItem('pmpro_mappingjson');
			window.location.href = window.location.href;
		});

		/*Create droppable element*/
		function init() {
			$('.makeMeDroppable').droppable({
				drop: handleDropEvent,
				hoverClass: 'hoverActive',
			});
			$('.pmpro-discord-roles-col').droppable({
				drop: handlePreviousDropEvent,
				hoverClass: 'hoverActive',
			});
		}

		$(init);

		/*Create draggable element*/
		function makeDrag(el) {
			// Pass me an object, and I will make it draggable
			el.draggable({
				revert: "invalid",
				helper: 'clone',
				start: function(e, ui) {
				ui.helper.css({"width":"45%"});
				}
			});
		}

		/*Handel droppable event for saved mapping*/
		function handlePreviousDropEvent(event, ui) {
			var draggable = ui.draggable;
			if(draggable.data('pmpro_level_id')){
				$(ui.draggable).remove().hide();
			}
			$(this).append(draggable);
			$('*[data-drop-pmpro_role_id="' + draggable.data('pmpro_role_id') + '"]').droppable({
				drop: handleDropEvent,
				hoverClass: 'hoverActive',
			});
			$('*[data-drop-pmpro_role_id="' + draggable.data('pmpro_role_id') + '"]').attr('data-drop-pmpro_role_id', '');

			var oldItems = JSON.parse(localStorage.getItem('pmpro_mapArray')) || [];
			$.each(oldItems, function (key, val) {
				if (val) {
					var arrayofval = val.split(',');
					if (arrayofval[0] == 'pmpro_level_id_' + draggable.data('pmpro_level_id') && arrayofval[1] == draggable.data('pmpro_role_id')) {
						delete oldItems[key];
					}
				}
			});
			var jsonStart = "{";
			$.each(oldItems, function (key, val) {
				if (val) {
					var arrayofval = val.split(',');
					if (arrayofval[0] != 'pmpro_level_id_' + draggable.data('pmpro_level_id') || arrayofval[1] != draggable.data('pmpro_role_id')) {
						jsonStart = jsonStart + '"' + arrayofval[0] + '":' + '"' + arrayofval[1] + '",';
					}
				}
			});
			localStorage.setItem('pmpro_mapArray', JSON.stringify(oldItems));
			var lastChar = jsonStart.slice(-1);
			if (lastChar == ',') {
				jsonStart = jsonStart.slice(0, -1);
			}

			var pmpro_mappingjson = jsonStart + '}';
			$("#pmpro_maaping_json_val").html(pmpro_mappingjson);
			localStorage.setItem('pmpro_mappingjson', pmpro_mappingjson);
			draggable.css({ 'width': '100%', 'left': '0', 'top': '0', 'margin-bottom': '10px' });
		}

		/*Handel droppable area for current mapping*/
		function handleDropEvent(event, ui) {
			var draggable = ui.draggable;
			var newItem = [];
			var newClone = $(ui.helper).clone();
			if($(this).find(".makeMeDraggable").length >= 1){
				return false;
			}
			$('*[data-drop-pmpro_role_id="' + newClone.data('pmpro_role_id') + '"]').droppable({
				drop: handleDropEvent,
				hoverClass: 'hoverActive',
			});
			$('*[data-drop-pmpro_role_id="' + newClone.data('pmpro_role_id') + '"]').attr('data-drop-pmpro_role_id', '');
			if ($(this).data('drop-pmpro_role_id') != newClone.data('pmpro_role_id')) {
				var oldItems = JSON.parse(localStorage.getItem('pmpro_mapArray')) || [];
				$(this).attr('data-drop-pmpro_role_id', newClone.data('pmpro_role_id'));
				newClone.attr('data-pmpro_level_id', $(this).data('pmpro_level_id'));

				$.each(oldItems, function (key, val) {
					if (val) {
						var arrayofval = val.split(',');
						if (arrayofval[0] == 'pmpro_level_id_' + $(this).data('pmpro_level_id')) {
							delete oldItems[key];
						}
					}
				});

				var newkey = 'pmpro_level_id_' + $(this).data('pmpro_level_id');
				oldItems.push(newkey + ',' + newClone.data('pmpro_role_id'));
				var jsonStart = "{";
				$.each(oldItems, function (key, val) {
					if (val) {
						var arrayofval = val.split(',');
						if (arrayofval[0] == 'pmpro_level_id_' + $(this).data('pmpro_level_id') || arrayofval[1] != newClone.data('pmpro_role_id') && arrayofval[0] != 'pmpro_level_id_' + $(this).data('pmpro_level_id') || arrayofval[1] == newClone.data('pmpro_role_id')) {
							jsonStart = jsonStart + '"' + arrayofval[0] + '":' + '"' + arrayofval[1] + '",';
						}
					}
				});

				localStorage.setItem('pmpro_mapArray', JSON.stringify(oldItems));
				var lastChar = jsonStart.slice(-1);
				if (lastChar == ',') {
					jsonStart = jsonStart.slice(0, -1);
				}

				var pmpro_mappingjson = jsonStart + '}';
				localStorage.setItem('pmpro_mappingjson', pmpro_mappingjson);
				$("#pmpro_maaping_json_val").html(pmpro_mappingjson);
			}

			// $(this).append(ui.draggable);
			// $(this).find('span').css({ 'order': '2' });
			$(this).append(newClone);
			$(this).find('span').css({ 'order': '2' });
			if (jQuery(this).find('.makeMeDraggable').length >= 1) {
				$(this).droppable("destroy");
			}
			makeDrag($('.makeMeDraggable'));
			newClone.css({ 'width': '100%', 'left': '0', 'top': '0', 'margin-bottom': '0px', 'position':'unset', 'order': '1' });
		}
	}

	/*Call-back on disconnect from discord*/
	$('#pmpro-disconnect-discord').on('click', function (e) {
		e.preventDefault();
		var userId = $(this).data('user-id');
		$.ajax({
			type: "POST",
			dataType: "JSON",
			url: etsPmproParams.admin_ajax,
			data: { 'action': 'disconnect_from_discord', 'user_id': userId, 'ets_discord_nonce': etsPmproParams.ets_discord_nonce, },
			beforeSend: function () {
				$(".ets-spinner").addClass("ets-is-active");
			},
			success: function (response) {
				if (response.status == 1) {
					window.location = window.location.href.split("?")[0];
				}
			},
			error: function (response) {
				console.error(response);
			}
		});
	});
	/*Call-back to manage member connection with discord from pmpro members-list*/
	$('.ets-run-api').on('click', function (e) {
		e.preventDefault();
		var userId = $(this).data('uid');
		$.ajax({
			type: "POST",
			dataType: "JSON",
			url: etsPmproParams.admin_ajax,
			data: { 'action': 'ets_pmpro_discord_member_table_run_api', 'user_id': userId, 'ets_discord_nonce': etsPmproParams.ets_discord_nonce, },
			beforeSend: function () {
				$("." + userId + ".spinner").addClass("is-active").show();
			},
			success: function (response) {
				if (response.status == 1) {
					$("." + userId + ".ets-save-success").show();;
				}
			},
			error: function (response) {
				console.error(response);
			},
			complete: function () {
				$("." + userId + ".spinner").removeClass("is-active").hide();
			}
		});
	});
	if (etsPmproParams.is_admin) {  
        $('#ets_pmpro_btn_color').wpColorPicker();
        $('#ets_pmpro_btn_disconnect_color').wpColorPicker();
    }
});
if (etsPmproParams.is_admin) {
	/*Tab options*/
	jQuery.skeletabs.setDefaults({
		keyboard: false,
	});
}